tooltips for node graphs

jseldess opened the pull request:
Based on session with @cuongdo, @MattTracy, and @maxlang, I've added basic tooltips for graphs on Nodes screen, both across all nodes and for a single node. 

I didn't add tooltips for "Internals" graphs that are exclusively for advanced internal debugging. Also, given space constraints, I didn't add much guidance in terms of what users should expect, troubleshooting, etc..

@spencerkimball, @petermattis, @bdarnell, would like your thoughts, too.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7087)

<!-- Reviewable:end -->

maxlang commented on ui/ts/pages/nodes.ts:
```diff
@@ -234,6 +234,7 @@ module AdminViews {
               Metrics.Select.Avg(_nodeMetric("sql.conns"))
                 .title("Client Connections")
             ).format(d3.format(".1")).title("SQL Connections")
+              .tooltip("The total number of active SQL connections to the cluster.")            
```
Do we want to use the "across all nodes" language here?

spencerkimball replied:
s/SQL connections/connections from SQL clients/

maxlang commented:
I'll port these into the new UI if people think they look good.

LGTM

spencerkimball commented on ui/ts/pages/nodes.ts:
```diff
@@ -245,6 +246,7 @@ module AdminViews {
                 .nonNegativeRate()
                 .title("Bytes Out")
             ).format(Utils.Format.Bytes).title("SQL Traffic")
+              .tooltip("The amount of network traffic sent to and from the SQL system, in bytes.")
```
s/network traffic sent to and from the SQL system/SQL client network traffic/

spencerkimball commented on ui/ts/pages/nodes.ts:
```diff
@@ -253,13 +255,16 @@ module AdminViews {
                 .title("Live Bytes")
-              ).format(Utils.Format.Bytes));
+              ).format(Utils.Format.Bytes)
+                .tooltip('The amount of storage space used by live (non-historical) CockroachDB data across all nodes.')
```
s/CockroachDB //

spencerkimball commented on ui/ts/pages/cluster.ts:
```diff
@@ -180,7 +180,7 @@ module AdminViews {
                 .title("Bytes Out")
             ).format(Utils.Format.Bytes).title("SQL Traffic")
-            .tooltip("The amount of network traffic sent to and from the SQL system, in bytes.")
+              .tooltip("The amount of network traffic sent to and from the SQL system, in bytes.")
```
s/network traffic sent to and from the SQL system/SQL client network traffic/

spencerkimball commented on ui/ts/pages/nodes.ts:
```diff
@@ -322,6 +330,7 @@ module AdminViews {
                 .nonNegativeRate()
                 .title("DDL Statements")
             ).format(d3.format(".1")).title("Schema Changes")
+              .tooltip("The number of DDL statements across all nodes, averaged over a 10 second period.")
```
What happens to these tooltips when the time interval is changed? Seems we should remove the "over a 10 second period" suffix and add it dynamically based on the current time interval.

mrtracy replied:
Max or I will add this once we actually do down-sample based on the time span; right now, all time spans are rendered with datapoints in 10 second intervals.

spencerkimball commented on ui/ts/pages/nodes.ts:
```diff
@@ -333,24 +342,27 @@ module AdminViews {
               Metrics.Select.Avg(_sysMetric("cpu.sys.percent"))
                 .title("CPU Sys %")
             ).format(d3.format(".2%")).title("CPU Usage").stacked(true)
+              .tooltip("The percentage of CPU used by CockroachDB (User %) and system-level operations (Sys %) across all nodes.")
```
Also, instead of always writing "across all nodes", let's define the base tooltip messages as constants somewhere and pull them in for the node and cluster pages and format each with the appropriate scope ("across all nodes", "on this node", etc.) and time interval ("over a 10 second period", "over a 1 hour period", etc.).

mrtracy replied:
I'll log an issue to look into this.

spencerkimball commented on ui/ts/pages/nodes.ts:
```diff
@@ -333,24 +342,27 @@ module AdminViews {
               Metrics.Select.Avg(_sysMetric("rss"))
                 .title("RSS")
             ).format(Utils.Format.Bytes).title("Memory Usage")
+              .tooltip("The memory in use across all nodes, broken down by CockroachDB Go code, the Go Runtime, and the total memory used by CockroachDB, including the key-value layer (RSS)."
```
s/CockroachDB Go code/CockroachDB Go allocations/

spencerkimball commented on ui/ts/pages/nodes.ts:
```diff
@@ -359,6 +371,7 @@ module AdminViews {
                 .nonNegativeRate()
                 .title("cgo Calls")
             ).format(d3.format(".1")).title("cgo Calls")
+              .tootlip("The number of calls to the C layer of CockroachDB across all nodes.")
```
s/to the C layer of CockroachDB/from Go to C by CockroachDB/

spencerkimball commented:
LGTM

jseldess commented:
*[ui/ts/pages/nodes.ts, line 266 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7087#-KJg-6S-LTmN0zl3FkAe-r2-266:-KJkSUcFepFPKZpUFtQB:169362333) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdaf7eae1e80fc41004d0210d9aa9e1ce52c629c/ui/ts/pages/nodes.ts#L266)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> s/CockroachDB //
</details>
Not sure what change your suggesting here, Spencer. Can you you write it out?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7087)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 0 of 2 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[ui/ts/pages/nodes.ts, line 266 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7087#-KJg-6S-LTmN0zl3FkAe-r2-266:-KJkTfDP2mq9ZsHLK0Zh:-1079378226) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdaf7eae1e80fc41004d0210d9aa9e1ce52c629c/ui/ts/pages/nodes.ts#L266)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> Not sure what change your suggesting here, Spencer. Can you you write it out?
</details>
Just doesn't seem useful to include the word CockroachDB here as it's pretty well-assumed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7087)*
<!-- Sent from Reviewable.io -->

jseldess commented:
*[ui/ts/pages/nodes.ts, line 333 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7087#-KJg-6S-LTmN0zl3FkAe-r2-333:-KJkTAIRhPRKqGVv0Iqv:326022095) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdaf7eae1e80fc41004d0210d9aa9e1ce52c629c/ui/ts/pages/nodes.ts#L333)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> What happens to these tooltips when the time interval is changed? Seems we should remove the "over a 10 second period" suffix and add it dynamically based on the current time interval.
</details>
This is a good catch. Here's what I see in the admin ui for the beta cluster:
- 10 min timescale = 10 second intervals
- 1 hour timescale = 20 second intervals BUT this seems inconsistent. Mostly, I see 20 second intervals. Sometimes, though, I see 10 second intervals. Matt, Max? 

For the 6 hours, 12 hours, and 1 day timescales, the admin UI lags too much for me to figure out what the reporting intervals are. Matt, Max, can you tell us the intervals? And is the lagging a known issue?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7087)*
<!-- Sent from Reviewable.io -->

jseldess commented:
Review status: 0 of 2 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[ui/ts/pages/cluster.ts, line 183 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7087#-KJg-6S-LTmN0zl3FkAd-r2-183:-KJkRvjD1ODyEGjZuNsL:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdaf7eae1e80fc41004d0210d9aa9e1ce52c629c/ui/ts/pages/cluster.ts#L183)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> s/network traffic sent to and from the SQL system/SQL client network traffic/
</details>
Done.

---

*[ui/ts/pages/nodes.ts, line 237 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7087#-KJg-6S-LTmN0zl3FkAe-r2-237:-KJkXBWy5ODCsiL8_5vR:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdaf7eae1e80fc41004d0210d9aa9e1ce52c629c/ui/ts/pages/nodes.ts#L237)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> s/SQL connections/connections from SQL clients/
> 
> s/to the cluster/across all nodes/ (be consistent – either use "all nodes" or "to the cluster" everywhere)
</details>
Done.

---

*[ui/ts/pages/nodes.ts, line 249 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7087#-KJg-6S-LTmN0zl3FkAe-r2-249:-KJkS6P9BPQdb-5ssc3U:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdaf7eae1e80fc41004d0210d9aa9e1ce52c629c/ui/ts/pages/nodes.ts#L249)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> s/network traffic sent to and from the SQL system/SQL client network traffic/
</details>
Done.

---

*[ui/ts/pages/nodes.ts, line 357 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7087#-KJg-6S-LTmN0zl3FkAe-r2-357:-KJkWPj-S0sq0oWSv0Oi:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdaf7eae1e80fc41004d0210d9aa9e1ce52c629c/ui/ts/pages/nodes.ts#L357)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> s/CockroachDB Go code/CockroachDB Go allocations/
</details>
Done.

---

*[ui/ts/pages/nodes.ts, line 374 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7087#-KJg-6S-LTmN0zl3FkAe-r2-374:-KJkWf6IMHZVb8T-rZrt:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdaf7eae1e80fc41004d0210d9aa9e1ce52c629c/ui/ts/pages/nodes.ts#L374)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> s/to the C layer of CockroachDB/from Go to C by CockroachDB/
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7087)*
<!-- Sent from Reviewable.io -->

jseldess commented:
Review status: 0 of 2 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[ui/ts/pages/nodes.ts, line 266 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7087#-KJg-6S-LTmN0zl3FkAe-r2-266:-KJkY17lmGVqr5tPOyUl:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdaf7eae1e80fc41004d0210d9aa9e1ce52c629c/ui/ts/pages/nodes.ts#L266)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Just doesn't seem useful to include the word CockroachDB here as it's pretty well-assumed.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7087)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
If you want to add this to the new UI, you also need to compile it by running `make` in the ui directory.  This will compile the typescript code and generate a new `embedded.go` file to check in.

mrtracy commented on ui/ts/pages/nodes.ts:
```diff
@@ -923,6 +949,7 @@ module AdminViews {
                 .sources([this._nodeId])
                 .title("goroutine count")
             ).format(d3.format(".1")).title("goroutine Count")
+              .tootlip("The number of Goroutines on this nodes. This count should rise and fall based on load.")
```
s/tootlip/tooltip/g

jseldess commented:
Review status: 0 of 2 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[ui/ts/pages/nodes.ts, line 952 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7087#-KJg-6S-LTmN0zl3FkAe-r3-952:-KJl9oAtxjhW0t52meIj:462386343) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fd674bf61feee189d5b9acab026d7a0e55acf98/ui/ts/pages/nodes.ts#L952)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> s/tootlip/tooltip/g
</details>
Done. Thanks for catching this. There were a few others as well. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7087)*
<!-- Sent from Reviewable.io -->

jseldess commented:
For those graphs that aren't showing snapshots in time, @nvanbenschoten pointed out that we should clarify that each data point is really the "average per second." For example, "The average number of SELECT statements per second across all nodes", not "The number of SELECT statements across all nodes, averaged over the last 10 seconds." 

I updated the text accordingly. @mrtracy, @spencerkimball, take one more look? Once we have more than just 10 second intervals, we could either add that detail back into the tooltips or, preferably, communicate this on screen when selecting the timespan.

spencerkimball commented on ui/ts/pages/cluster.ts:
```diff
@@ -180,16 +179,16 @@ module AdminViews {
                 .title("Bytes Out")
             ).format(Utils.Format.Bytes).title("SQL Traffic")
-            .tooltip("The amount of network traffic sent to and from the SQL system, in bytes.")
+              .tooltip("The average amount of SQL client network traffic per second.")
```
s/second/second, in bytes/

spencerkimball commented on ui/ts/pages/nodes.ts:
```diff
@@ -245,6 +246,7 @@ module AdminViews {
                 .nonNegativeRate()
                 .title("Bytes Out")
             ).format(Utils.Format.Bytes).title("SQL Traffic")
+              .tooltip("The average amount of SQL client network traffic per second across all nodes.")
```
s/per second/in bytes per second/

spencerkimball commented:
LGTM

Commit:
Add Tooltips to Node/Nodes Graph Pages

This has only been done in the current ui; it must be ported to ui/next in a
subsequent commit.

mberhault commented on ui/ts/pages/cluster.ts:
```diff
@@ -158,16 +158,15 @@ module AdminViews {
             ).format(Utils.Format.Bytes).title("Memory Usage")
-              .tooltip("The memory in use across all nodes.")
-
+              .tooltip("The average memory in use across all nodes.")
```
I think I just changed that in my PR. It's definitely not average, it's the total memory used.

mberhault commented on ui/ts/pages/nodes.ts:
```diff
@@ -348,13 +358,15 @@ module AdminViews {
               Metrics.Select.Avg(_sysMetric("rss"))
                 .title("RSS")
             ).format(Utils.Format.Bytes).title("Memory Usage")
+              .tooltip("The memory in use across all nodes, broken down by CockroachDB Go allocations, the Go Runtime, and the total memory used by CockroachDB, including the key-value layer (RSS).")
```
ooph. this is not very clear.
I'm not sure what the best way to put it is, but here's a summary of each of the metrics:
* rss: total memory used by the cockroach node
* go alloc: memory currently allocated by the Go layer
* go total: total memory managed by the Go allocator (includes metadata and idle memory)
* cgo alloc: memory currently allocated by the C++ layer
* go total: total memory managed by the C++ allocator (includes metadata and idle memory)
